relm4-examples/libadwaita/examples/view-switcher.rs

```diff
@@ -0,0 +1,78 @@
+use adw::traits::ApplicationWindowExt;
+use gtk::prelude::{BoxExt, ButtonExt, GtkWindowExt, OrientableExt};
+use relm4::{send, AppUpdate, Model, RelmApp, Sender, WidgetPlus, Widgets};
+
+#[derive(Default)]
+struct AppModel {
+    counter: u8,
+}
+
+enum AppMsg {
+    Increment,
+    Decrement,
+}
+
+impl Model for AppModel {
+    type Msg = AppMsg;
+    type Widgets = AppWidgets;
+    type Components = ();
+}
+
+impl AppUpdate for AppModel {
+    fn update(&mut self, msg: AppMsg, _components: &(), _sender: Sender<AppMsg>) -> bool {
+        match msg {
+            AppMsg::Increment => {
+                self.counter = self.counter.wrapping_add(1);
+            }
+            AppMsg::Decrement => {
+                self.counter = self.counter.wrapping_sub(1);
+            }
+        }
+        true
+    }
+}
+
+fn application_window() -> adw::ApplicationWindow {
+    adw::ApplicationWindow::builder().build()
+}
+
+#[relm4_macros::widget]
+impl Widgets<AppModel, ()> for AppWidgets {
+    view! {
+        main_window = application_window() -> adw::ApplicationWindow {
+            set_default_width: 300,
+            set_default_height: 200,
+
+            set_content = Some(&gtk::Box) {
+                set_orientation: gtk::Orientation::Vertical,
+                set_margin_all: 5,
+                set_spacing: 5,
+
+                append = &adw::HeaderBar {
+                    set_title_widget = Some(&adw::ViewSwitcherTitle) {
+                        set_title: "A stack switcher App",
+                        set_stack: Some(&stack),
+                    }
+                },
+                
+                append: stack = &adw::ViewStack {
+                    add_titled(Some("First"), "First Page") = &gtk::Label {
+                        set_label: "This is the start page",
```

Comment:
Sorry, I'm stuck. What would be the way to add icons and such?
To me the way how to untilize [set_icon_name](https://world.pages.gitlab.gnome.org/Rust/libadwaita-rs/stable/latest/docs/libadwaita/struct.ViewStackPage.html#method.set_icon_name) is still mysterious

Reply:
Before using set_icon_name you either have to know how your theme calls a certain icon (which is not very reliable) or you add your own image as resource (there's an "Icon library" app to help you btw.). Using your own resources is a bit painful though because it usually requires meson as build system and a resource schema to load the image. It think you can also use resources with cargo and pure Rust but I haven't tried it yet.

An example code to load a resource files can be found here: https://gitlab.elektrotechnik.hs-augsburg.de/erhardt/state-factory/-/blob/gtk4/src/static_resources.rs (lines 22-24)

And together with this function you might be able to get images working without meson: https://gtk-rs.org/gtk-rs-core/git/docs/gio/fn.compile_resources.html

My idea would be: compile your XML schema (schemas look like [this](https://gitlab.elektrotechnik.hs-augsburg.de/erhardt/state-factory/-/blob/gtk4/data/resources/resources.gresource.xml)) -> load compiled schema -> register schema for your application -> use icon

Reply:
Probably I should have been more precise with my problem description. I understand the concept with the named icons. The problem I have here is, how to create a `libadwaita::ViewStackPage` as it does not seam to have a constructor and somehow just gets implicitly created by using e.g. `add_titled`. But how can I later add  those icons to it?

Reply:
An other question I have: Can I also combine `property_name(value1, value2, ...) = gtk::Box { ... }` with providing a name `property_name: name = gtk::Box { ... }`?
Maybe something like `property_name: name(value1, value2, ...) = gtk::Box { ... }`

Reply:
The problem is that Relm4 doesn't handle cases yet where a property function returns a new widget. Libadwaita differs from GTK in this point and the macro will probably need a new syntax for this.

An yes you should be able to use a name for the widget. I think the syntax is like this:
`property(arg1, arg2, ...): name = gtk::Box { ... }`

Reply:
Would it make sense to make them aware of this problem?
Maybe they could just add additional methods to avoid that problem. Or do you think it's not worth the effort?

Reply:
I don't think they would introduce a new API for us unless it males sense in general. So if you have an idea for a new ViewStack API an issue would make sense. Otherwise Relm4 needs to adapt, I guess.

Reply:
I would think about something like this
``` rust
                append: stack = &adw::ViewStack {
                    add: &adw::ViewStackPage {
                        set_child: = &gtk::Label {
                            set_label: "This is the start page",
                        },
                        set_icon_name: Some("icon_name"),
                        set_badge_number: 3,
                    },
                    add_titled(Some("First"), "First Page") = &gtk::Label {
                        set_label: "This is the start page",
                    },
```
What do you think about that? This would probably need two additional methods:

- `adw::ViewStackPage::new()`
- `adw::ViewStackPage::set_child(&self, child: &P)`

Reply:
That's what I'd suggest as well. We need some sort of constructor and a method to set the child. Or even both at once. Do you have time to open an issue over at libadwaita?

Reply:
Yes, currently I have time, but not sure, how the C-API looks like :smile: 
I'll will have a look.

Reply:
You're basically looking for a new `view_stack_page_new` and `view_stack_page_set_child`method when "talking C".
More here: https://gnome.pages.gitlab.gnome.org/libadwaita/doc/1.0.0.alpha.4/

Reply:
What do you think? https://gitlab.gnome.org/GNOME/libadwaita/-/issues/311

Reply:
Looks good. Let's see what they think :)

Reply:
Thanks for opening the issue! I already expected they wouldn't be happy about these changes (and to be honest I wouldn't be either in their place). That then means more work for me and a new syntax for handling these types of functions... suggestions are welcome :)